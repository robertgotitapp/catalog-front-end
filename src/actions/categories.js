import { CategoriesAction } from '../utils/const';
import { post, get } from '../utils/requests';

export function addCategory(accessToken, category) {
  return {
    type: CategoriesAction.ADD_CATEGORY,
    promise: post('http://127.0.0.1:5000/categories', accessToken, category),
  };
}

export function getCategories(offset, limit) {
  return {
    type: CategoriesAction.GET_CATEGORIES,
    promise: get(`http://127.0.0.1:5000/categories?offset=${offset}&limit=${limit}`),
  };
}

export function selectCurrentCategory(selectedCategory) {
  return {
    type: CategoriesAction.SELECT_CURRENT_CATEGORY,
    payload: { currentCategory: selectedCategory },
  };
}
