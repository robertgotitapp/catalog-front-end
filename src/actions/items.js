import { ItemsAction, HeadersType } from '../utils/const';
import {
  get, post, update, remove,
} from '../utils/requests';

export function addItem(categoryId, item) {
  return {
    type: ItemsAction.ADD_ITEM,
    promise: post(`http://127.0.0.1:5000/categories/${categoryId}/items`,
      [HeadersType.CONTENTTYPE, HeadersType.AUTHORIZATION],
      item),
  };
}

export function getItems(categoryId, offset, limit) {
  return {
    type: ItemsAction.GET_ITEMS,
    promise: get(`http://127.0.0.1:5000/categories/${categoryId}/items?offset=${offset}&limit=${limit}`),
  };
}

export function removeItem(categoryId, itemId) {
  return {
    type: ItemsAction.DELETE_ITEM,
    promise: remove(`http://127.0.0.1:5000/categories/${categoryId}/items/${itemId}`,
      [HeadersType.CONTENTTYPE, HeadersType.AUTHORIZATION]),
  };
}

export function updateItem(categoryId, itemId, item) {
  console.log(categoryId);
  console.log(item);
  return {
    type: ItemsAction.UPDATE_ITEM,
    promise: update(
      `http://127.0.0.1:5000/categories/${categoryId}/items/${itemId}`,
      [HeadersType.CONTENTTYPE, HeadersType.AUTHORIZATION],
      item,
    ),
  };
}
