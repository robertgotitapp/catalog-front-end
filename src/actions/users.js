export const ADD_USER = 'ADD_USER'
export const AUTHENTICATE_USER = 'AUTHENTICATE_USER'

export function addUser(user) {
    return {
        type: ADD_USER,
        user
    }
}

export function handleAddUser(user) {
    return (dispatch) => {
        dispatch(addUser(user))
    }
}

