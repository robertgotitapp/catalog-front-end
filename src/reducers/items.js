import { ItemsAction } from '../utils/const';

const initialState = { items: {}, currentPage: null };

export default function items(state = initialState, action) {
  switch (action.type) {
    case ItemsAction.GET_ITEMS_SUCCESS:
      return {
        ...state,
        items: action.payload.items,
        totalItems: action.payload.total_items,
      };
    case ItemsAction.ADD_ITEM_SUCCESS:
      return {
        ...state,
        [action.payload.id]: action.payload,
      };
    case ItemsAction.UPDATE_ITEM_SUCCESS:
      return {
        ...state,
        [action.payload.id]: action.payload,
      };
    case ItemsAction.SELECT_ITEM_PAGE:
      return {
        ...state,
        currentPage: action.payload.currentPage,
      };
    default:
      return state;
  }
}
