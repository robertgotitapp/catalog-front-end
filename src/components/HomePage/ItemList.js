import React, { Component } from 'react';
import { connect } from 'react-redux';
import Card from 'react-bootstrap/Card';
import { Link } from 'react-router-dom';
import Pagination from 'react-bootstrap/Pagination';
import Item from './Item';
import { getItems } from '../../actions/items';
import { PaginationConfig } from '../../utils/const';

export class ItemList extends Component {
  goToPage = (destinationPage) => {
    const offset = (destinationPage - 1) * PaginationConfig.ITEMS_PER_PAGE;
    this.props.getItems(this.props.categoryId, offset, PaginationConfig.ITEMS_PER_PAGE);
    this.props.history.push(`/categories/${this.props.categoryId}/items/${destinationPage}`);
  }

  render() {
    return (
      <div>
        <Card>
          <Card.Body>
            <Card.Title>Items</Card.Title>
          </Card.Body>
          <div>
            { this.props.itemIds.map(id => (
              <Link className='item-link item-card' key={id} to={`/items/${this.props.items[id].id}`}>
                <Item id={id} />
              </Link>
            ))
          }
          </div>
          <Pagination>
            {
              // Current Page does not have onClick event but have active attribute
              // and likewise.
              this.props.pageList.map(pageNumber => (
                Number(pageNumber) === Number(this.props.pageNumber)
                  ? (
                    <Pagination.Item
                      key={pageNumber}
                      name={pageNumber}
                      active
                    >
                      {pageNumber}
                    </Pagination.Item>
                  )
                  : (
                    <Pagination.Item
                      key={pageNumber}
                      name={pageNumber}
                      className='clickablePage'
                      onClick={() => this.goToPage(pageNumber)}
                    >
                      {pageNumber}
                    </Pagination.Item>
                  )
              ))
            }
          </Pagination>
        </Card>
      </div>
    );
  }
}

function mapStateToProps({ items }, { pageNumber }) {
  // Find out the right most page and left most page being display
  // in the pagination, then create an array containing list of
  // those pages being displayed
  const lastPage = Math.floor((items.totalItems - 1) / 10) + 1;
  const leftPage = Math.max(pageNumber - 2, 1);
  const rightPage = Math.min(lastPage, pageNumber + 2);
  const pageList = [];
  for (let i = leftPage; i <= rightPage; i += 1) {
    pageList.push(i);
  }
  return {
    itemIds: Object.keys(items.items),
    items: items.items,
    pageNumber,
    pageList,
  };
}

const mapDispatchToProps = {
  getItems,
};

export default connect(mapStateToProps, mapDispatchToProps)(ItemList);
