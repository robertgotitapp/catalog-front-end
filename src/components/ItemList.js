import React, { Component } from 'react';
import { connect } from 'react-redux';
import Card from 'react-bootstrap/Card';
import { Link } from 'react-router-dom';
import Pagination from 'react-bootstrap/Pagination';
import Item from './Item';
import { getItems, selectItemPage } from '../actions/items';

export class ItemList extends Component {
  goToPage = (e) => {
    e.preventDefault();
    if (e.target.innerHTML) {
      const destinationPage = Number(e.target.innerHTML);
      this.props.selectItemPage(destinationPage);
      const offset = (destinationPage - 1) * 10;
      const limit = 10;
      this.props.getItems(this.props.selectedCategory.id, offset, limit);
    }
  }

  render() {
    return (
      <div>
        <Card>
          <Card.Body>
            <Card.Title>Items</Card.Title>
          </Card.Body>
          { this.props.itemIds.map(id => (
            <Link key={id} to={`/items/${this.props.items[id].id}`}>
              <Item id={id} />
            </Link>
          ))
          }
          <Pagination>
            {
              // Current Page does not have onClick event but have active attribute
              // and likewise.
              this.props.pageList.map(pageNumber => (
                pageNumber === this.props.currentPage
                  ? (
                    <Pagination.Item
                      key={pageNumber}
                      active
                    >
                      {pageNumber}
                    </Pagination.Item>
                  )
                  : (
                    <Pagination.Item
                      key={pageNumber}
                      onClick={this.goToPage}
                    >
                      {pageNumber}
                    </Pagination.Item>
                  )
              ))
            }
          </Pagination>
        </Card>
      </div>
    );
  }
}

function mapStateToProps({ itemsReducer, categoriesReducer }) {
  const lastPage = Math.floor((itemsReducer.totalItems - 1) / 10) + 1;
  const { currentPage } = itemsReducer;
  const leftPage = Math.max(currentPage - 2, 1);
  const rightPage = Math.min(lastPage, currentPage + 2);
  const pageList = [];
  for (let i = leftPage; i <= rightPage; i += 1) {
    pageList.push(i);
  }
  return {
    itemIds: Object.keys(itemsReducer.items),
    items: itemsReducer.items,
    selectedCategory: categoriesReducer.currentCategory,
    currentPage,
    pageList,
  };
}

const mapDispatchToProps = {
  getItems,
  selectItemPage,
};

export default connect(mapStateToProps, mapDispatchToProps)(ItemList);
