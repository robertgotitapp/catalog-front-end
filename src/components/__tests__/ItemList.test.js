import React from 'react';
import { shallow } from 'enzyme';
import { Link } from 'react-router-dom';
import PageItem from 'react-bootstrap/PageItem';
import { ItemList } from '../ItemList';
import { getItems, selectItemPage } from '../../actions/items';

describe('components/ItemList', () => {
  const props = {
    items: {
      4: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 4,
        name: 'Item 4',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
      5: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 5,
        name: 'Item 4',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
      6: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 6,
        name: 'Item 6',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
      7: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 7,
        name: 'Item 4',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
      8: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 8,
        name: 'Item 4',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
      9: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 9,
        name: 'Item 4',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
      10: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 10,
        name: 'Item 4',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
      11: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 11,
        name: 'Item 4',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
      12: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 12,
        name: 'Item 4',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
      13: {
        category_id: 1,
        created: '2019-06-17T13:49:00+00:00',
        description: 'Item 4',
        id: 13,
        name: 'Item 4',
        price: 19,
        updated: '2019-06-17T13:49:00+00:00',
      },
    },
    itemIds: [4, 5, 6, 7, 8, 9, 10, 11, 12, 13],
    selectedCategory: {
      created: '2019-06-17T13:46:18+00:00',
      description: 'Laptop',
      id: 1,
      name: 'Laptop',
      updated: '2019-06-17T13:46:18+00:00',
    },
    currentPage: 1,
    pageList: [1, 2],
    getItems,
    selectItemPage,
  };
  const container = shallow(<ItemList {...props} />);
  container.setState({ loading: false });

  it('should render correctly', () => {
    expect(container).toMatchSnapshot();
  });

  it('should contain exact number of item links', () => {
    expect(container.find(Link)).toHaveLength(10);
  });

  it('should contain exact number of PageItem', () => {
    expect(container.find(PageItem)).toHaveLength(2);
  });

  it('should contain exact number of clickable link', () => {
    expect(container.find('[name="clickablePage"]')).toHaveLength(1);
  });

  it('should trigger goToPage handler when clickable page is clicked', () => {
    container.instance().goToPage = jest.fn();
    container.instance().forceUpdate();
    const form = container.find('[name="clickablePage"]');
    form.simulate('click');
    expect(container.instance().goToPage).toHaveBeenCalled();
  });
});
