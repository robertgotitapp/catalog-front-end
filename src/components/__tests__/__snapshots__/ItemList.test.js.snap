// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ItemList should render correctly 1`] = `
<div>
  <Card
    body={false}
  >
    <CardBody>
      <CardTitle>
        Items
      </CardTitle>
    </CardBody>
    <Link
      key="4"
      to="/items/4"
    >
      <Connect(Item)
        id={4}
      />
    </Link>
    <Link
      key="5"
      to="/items/5"
    >
      <Connect(Item)
        id={5}
      />
    </Link>
    <Link
      key="6"
      to="/items/6"
    >
      <Connect(Item)
        id={6}
      />
    </Link>
    <Link
      key="7"
      to="/items/7"
    >
      <Connect(Item)
        id={7}
      />
    </Link>
    <Link
      key="8"
      to="/items/8"
    >
      <Connect(Item)
        id={8}
      />
    </Link>
    <Link
      key="9"
      to="/items/9"
    >
      <Connect(Item)
        id={9}
      />
    </Link>
    <Link
      key="10"
      to="/items/10"
    >
      <Connect(Item)
        id={10}
      />
    </Link>
    <Link
      key="11"
      to="/items/11"
    >
      <Connect(Item)
        id={11}
      />
    </Link>
    <Link
      key="12"
      to="/items/12"
    >
      <Connect(Item)
        id={12}
      />
    </Link>
    <Link
      key="13"
      to="/items/13"
    >
      <Connect(Item)
        id={13}
      />
    </Link>
    <Bootstrap(Pagination)>
      <PageItem
        active={true}
        activeLabel="(current)"
        disabled={false}
        key="1"
      >
        1
      </PageItem>
      <PageItem
        active={false}
        activeLabel="(current)"
        disabled={false}
        key="2"
        name="clickablePage"
        onClick={[Function]}
      >
        2
      </PageItem>
    </Bootstrap(Pagination)>
  </Card>
</div>
`;
